src/Function.cpp

```diff
@@ -248,6 +299,129 @@ Function::Function(const std::string &n) : contents(new FunctionContents) {
     contents.ptr->name = n;
 }
 
+// Deep-copy UpdateDefinition from 'src' to 'dst'
+void deep_copy_update_definition_helper(UpdateDefinition &dst,
+                                        const UpdateDefinition &src,
+                                        DeepCopyMap &copied_map) {
+    dst.values = src.values;
+    dst.args = src.args;
+    dst.schedule = src.schedule.deep_copy(copied_map);
+
+    // UpdateDefinition's domain is the same as the one pointed by its schedule
+    internal_assert(src.schedule.reduction_domain().same_as(src.domain))
+        << "UpdateDefinition should point to the same reduction domain as its schedule\n";
+    // We don't need to deep-copy the reduction domain since we've already done
+    // it when deep-copying the schedule above
+    dst.domain = dst.schedule.reduction_domain();
+}
+
+// Deep-copy ExternFuncArgument from 'src' to 'dst'
+void deep_copy_extern_func_argument_helper(ExternFuncArgument &dst,
```

Comment:
The problem is that FunctionContents and ScheduleContents is defined as private structs inside the .cpp files. If we were to make DeepCopyMap into a separate class, we would have to defined it in multiple files. I changed deep_copy_extern_func_argument_helper and deep_copy_update_definition_helper to return value instead, maybe that would it less confusing? The deep_copy_function_contents_helper and deep_copy_schedule_contents_helper still needs to take 'dst' as input however since Function and Schedule automatically allocates a pointer for the contents.